### PR TITLE
Change the type of the receiver in type methods from _any to the defining type

### PR DESCRIPTION
**Pull request by daviditen**
Change the type of the receiver for type methods from _any to the type of the
class/record/union that defined the method.  This prevents calling type
methods on types other than the one that defined them and allows overloading
type methods.

Removed two .futures.

### CODE REVIEW

File: compiler/passes/cleanup.cpp [hunk 1]
```diff
@@ -206,6 +206,20 @@ static void add_parens_to_deinit_fns(FnSymbol* fn) {
...
 
+/* Change the type of the receiver in type methods from "_any" to
+   the type that defined the method.
+*/
+static void set_receiver_type_in_type_methods(FnSymbol* fn) {
+  if (fn->hasFlag(FLAG_METHOD) && fn->thisTag == INTENT_TYPE) {
+    if (TypeSymbol* ts = toTypeSymbol(fn->defPoint->parentSymbol)) {
+      if (fn->_this->type == dtAny) {
+        fn->_this->type = ts->type;
+      }
```
Comment (lydia-duncan) [inline compiler/passes/cleanup.cpp]: Does this code assume the type method is a primary method?  What happens if the type method is a secondary method?

Comment (lydia-duncan) [inline compiler/passes/cleanup.cpp]: As Mike and Michael will attest, I've been on a tear about secondary methods recently.  If it doesn't work, feel free to just add a future/issue for the secondary method case and move on.

**Comment by lydia-duncan** [comment]
Looks good to me, other than the comment.  Thanks!

File: compiler/passes/cleanup.cpp (hunk 1 above)
Comment (daviditen) [inline compiler/passes/cleanup.cpp]: Secondary methods fall through here, and basically the same transformation happens to them in normalization, though it wasn't super obvious exactly where in normalization it happens.

Comment (lydia-duncan) [inline compiler/passes/cleanup.cpp]: Cool!  Can you add a comment to that effect?

Comment (daviditen) [inline compiler/passes/cleanup.cpp]: Will do.  Thank you!

**Comment by noakesmichael** [comment]
I also worried about secondary methods and so I did a test on that case
and determined that they work fine on master.

Currently secondary methods have a typeExpr field inserted in the
ArgSymbol for fn->_this and, as David indicates, this is used to fix-up
the type of “this” somewhere in normalize().

Longer term it seems like a bug to me that primary methods and
second methods rely on different mechanisms to determine the
type of “this”.  I mean this in the general case and not merely for
type methods.

Mike




On Jan 11, 2017, at 2:24 PM, David Iten <notifications@github.com<mailto:notifications@github.com>> wrote:


@daviditen commented on this pull request.

________________________________

In compiler/passes/cleanup.cpp<https://github.com/chapel-lang/chapel/pull/5153>:

> @@ -206,6 +206,20 @@ static void add_parens_to_deinit_fns(FnSymbol* fn) {
 }


+/* Change the type of the receiver in type methods from "_any" to
+   the type that defined the method.
+*/
+static void set_receiver_type_in_type_methods(FnSymbol* fn) {
+  if (fn->hasFlag(FLAG_METHOD) && fn->thisTag == INTENT_TYPE) {
+    if (TypeSymbol* ts = toTypeSymbol(fn->defPoint->parentSymbol)) {
+      if (fn->_this->type == dtAny) {
+        fn->_this->type = ts->type;
+      }


Secondary methods fall through here, and basically the same transformation happens to them in normalization, though it wasn't super obvious exactly where in normalization it happens.

**Pull request merged**